### PR TITLE
Bug 1857246: Retrying in case of conflict during Config update

### PR DESCRIPTION
**Pull request by ricardomaraschini**
We have multiple controllers concurrently writting the status of the
config, this commit introduces retries in case of any of them updated
the config resource before us. This is clearly a workaround, as the
main controller may create storage if we fail to update the
configuration here it may ended up creating storage again on the next
cycle, hence the need to make sure we succeed.

### CODE REVIEW

**Comment by dmage** [comment]
@ricardomaraschini can you create an e2e test that checks that after the bootstrap storage is created only once?

**Comment by ricardomaraschini** [comment]
> @ricardomaraschini can you create an e2e test that checks that after the bootstrap storage is created only once?

As to have tests for every platform would be overkill I changed the error message and implemented a test to make sure it is not happening. PTAL.

File: test/e2e/bootstrap_test.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/openshift/cluster-image-registry-operator/test/framework"
+)
+
+func TestBootstrapFailToUpdateSpec(t *testing.T) {
+	te := framework.Setup(t)
```
Comment (dmage) [inline test/e2e/bootstrap_test.go:11]: Use SetupAvailableImageRegistry instead of this bunch of Deploy/Wait/Ensure calls.

**Comment by dmage** [comment]
/hold
it doesn't work

File: pkg/operator/controller.go [hunk 2]
```diff
@@ -247,14 +248,26 @@ func (c *Controller) sync() error {
...
-		)
-		if err != nil {
-			if !errors.IsConflict(err) {
-				klog.Errorf("unable to update %s: %s", utilObjectInfo(cr), err)
+		var updatedCR *imageregistryv1.Config
+		if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+			updatedCR, err = c.listers.RegistryConfigs.Get(defaults.ImageRegistryResourceName)
+			if err != nil {
+				return err
 			}
```
Comment (dmage) [inline pkg/operator/controller.go:256]: the object from the lister shouldn't be mutated, DeepCopy may help

Comment (ricardomaraschini) [reply pkg/operator/controller.go:256]: done.

**Pull request merged**